python/pyiceberg/expressions/literals.py

```diff
@@ -108,7 +110,7 @@ def __ge__(self, other):
 
 
 @singledispatch
-def literal(value) -> Literal:
+def literal(value: Any) -> Literal[T]:
```

Comment:
There are more: https://github.com/apache/iceberg/blob/master/python/pyiceberg/expressions/literals.py#L159-L161 For example here we convert the `bytearray` into `bytes`. Do you want to do this in this PR?

Reply:
Let me take a closer look. Maybe we should return `Literal[bytearray]`.

Reply:
We could also just accept bytes, and then it is up for the user to convert. I think that would be fine too. The difference is mutability: `bytes` is immutable, and `bytearray` is mutable.

Reply:
I think what worries me here is that I don't understand what `T` is in this case. Is this the equivalent in Java?

```java
  <T> Literal<T> of(Object value) {
    return new Literal<>(value);
  }
```

That basically lets you return the parameterized type even though it may not match because the type comes from the call context.

Reply:
Yes, that's the case for Python as well. I'm happy to fix this, but I'd rather do it in another PR since it is not really related to the dataclasses.